Add support for ILM transition.

poornas opened the pull request:
This PR adds transition support for ILM
to transition data to another MinIO target
represented by a storage class ARN. Subsequent
GET or HEAD for that object will be streamed from
the transition tier. If PostRestoreObject API is
invoked, the transitioned object can be restored for
duration specified to the source cluster.

## Description
Needs https://github.com/minio/minio-go/pull/1390 and https://github.com/minio/mc/pull/3420

## Motivation and Context
Allows applications to transition data to less expensive storage, e.g from NVME to hard disks. 

## How to test this PR?
Use companion PR(s) listed above. 
```
1. Set up a transition target for srcbucket on myminio to dstbucket on minio2:9000
mc admin bucket remote add  myminio/srcbucket \
        https://foobar:foo12345@minio2:9000/dstbucket \
        --service "ilm" --region "us-west-1" --label "transitiontier"

2. Use arn generated above to add a ilm rule with transition destination specified by the --storageclass flag
mc ilm add --expiry-days 3 --transition-days 2 --storage-class "transitiontier" source/srcbucket                         

When a object qualifies for transitioning per the rules set in the ilm config, it should be moved to the destination cluster. However all GET and HEAD should work. If post restore object API is called on a transitioned object, it should generate a local copy of object on source cluster.

For example:
aws s3api restore-object --endpoint-url http://localhost:9008 --profile minio \
--bucket srcbucket \
--key x2 \
--restore-request Days=1
```

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,464 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+		return false, "", BucketRemoteDestinationNotFound{Bucket: arn.Bucket}
+	}
+	// validate transition ARN against target endpoint
+	c, ok := globalBucketTargetSys.arnRemotesMap[arnStr]
```
maybe we should protect this access with a lock ?

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,464 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+	// validate transition ARN against target endpoint
+	c, ok := globalBucketTargetSys.arnRemotesMap[arnStr]
+	if ok {
+		if c.EndpointURL().String() == clnt.EndpointURL().String() {
```
I don't like this comparaison a lot since a trailing slash can make the comparaison wrong, maybe let's compare host & port

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,464 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+	var opts ObjectOptions
+	opts.Versioned = globalBucketVersioningSys.Enabled(bucket)
+	opts.VersionID = objInfo.VersionID
+	if time.Now().UTC().After(objInfo.Expires) {
```
I think objInfo.Expires is a totally different thing here, this is not related to lifecycle expiry

poornas replied:
thanks for catching that, fixed

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,464 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+
+// getTransitionedObjectReader returns a reader from the transitioned tier.
+func getTransitionedObjectReader(ctx context.Context, bucket, object string, rs *HTTPRangeSpec, h http.Header, oi ObjectInfo, opts ObjectOptions) (gr *GetObjectReader, err error) {
+	var lc *lifecycle.Lifecycle
```
Have difficulty to understand why lifecycle is needed here

poornas replied:
we need to get the transition storage class to identify the actual storage location

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,464 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+	if _, err := objAPI.PutObject(ctx, bucket, object, pReader, opts); err != nil {
+		return err
+	}
+	gr.Close()
```
Close with defer in the beginning of this function ?

vadmeste commented on cmd/data-crawler.go:
```diff
@@ -738,8 +745,28 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
 			}
 		}
 		opts.VersionID = obj.VersionID
+		if obj.Transitioned {
```
I think this code could be moved to little later, 
```
if obj.Transitioned {
   deleteTransitionedObject()
} else {
   o.DeleteObject()
}
```

poornas replied:
done

vadmeste commented on pkg/bucket/lifecycle/lifecycle.go:
```diff
@@ -275,3 +326,13 @@ func (lc Lifecycle) PredictExpiryTime(obj ObjectOpts) (string, time.Time) {
 }
+
+// GetTransitionARN returns transition ARN.
+func (lc Lifecycle) GetTransitionARN(obj ObjectOpts) string {
```
Can we call this GetTransitionStorageClass ? since we are in lifecycle package, it will be less confusing that way

poornas replied:
done

vadmeste commented on pkg/bucket/lifecycle/lifecycle.go:
```diff
@@ -160,15 +169,26 @@ func (lc Lifecycle) FilterActionableRules(obj ObjectOpts) []Rule {
+		// The NoncurrentVersionTransition action requests MinIO to transition
+		// noncurrent versions of objects x days after the objects become
+		// noncurrent.
+		if !rule.NoncurrentVersionTransition.IsDaysNull() {
```
@poornas the point of FilterActionableRules is return rules that match prefix & tags, rules that could be actionable. I know that function name is not good, but it is just easier move the decision outside this function

poornas replied:
transition rules can also come with prefix and tags - so made sense to do this check within this function

vadmeste reviewed (commented):
First round of review, some comments, not tested yet

harshavardhana commented on cmd/admin-bucket-handlers.go:
```diff
@@ -135,10 +135,6 @@ func (a adminAPIHandlers) SetRemoteTargetHandler(w http.ResponseWriter, r *http.
 		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrServerNotInitialized), r.URL)
 		return
 	}
-	if !globalIsErasure {
-		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrNotImplemented), r.URL)
-		return
-	}
```
Any reason these are removed?

poornas replied:
Lifecycle transition is permitted in FS mode, hence this check is moved to SetTarget as a conditional check for replication.

harshavardhana commented on cmd/admin-bucket-handlers.go:
```diff
@@ -261,10 +257,6 @@ func (a adminAPIHandlers) RemoveRemoteTargetHandler(w http.ResponseWriter, r *ht
 		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrServerNotInitialized), r.URL)
 		return
 	}
-	if !globalIsErasure {
-		writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrNotImplemented), r.URL)
-		return
-	}
```
Any reason these are removed?

poornas replied:
same reasoning as above.

harshavardhana commented on cmd/admin-router.go:
```diff
@@ -181,7 +181,9 @@ func registerAdminRouter(router *mux.Router, enableConfigOps, enableIAMOps bool)
-			// Bucket replication operations
+		}
+		if env.Get(envDataUsageCrawlConf, config.EnableOn) == config.EnableOn {
+			// Bucket replication/lifecycle operations
```
Because the crawler runs for FS mode as well also no need to extend storage tiering to FS mode. I would say lets not increase the scope for single drive mode.

poornas replied:
ok, will confirm with AB whether we can drop tiering in FS mode, then update the PR accordingly

harshavardhana replied:
IMO FS mode should be left alone because of complications this can bring in, because we don't do replication while we still have crawler in FS mode. 

There are no serious deployments in FS mode which warrant this - it simply makes it complicated.

harshavardhana replied:
You will have to explain why versioning won't work with ILM transition etc in FS mode.

poornas replied:
just checked with AB, he thinks it makes sense to support transition in FS mode since ILM is already a supported feature.His suggestion was to separate into 2 PR's if that would make it easier.

harshavardhana replied:
My problem is increasing the scope of FS mode @poornas it really doesn't make sense to me. While we support ILM in FS we don't support all features of ILM i.e DeleteMarker, NoCurrentVersion etc.

No one will ever use it and is asking for it. I would say the first step add this for erasure coding and stabilize it and add it for FS mode if needed in the future.

poornas replied:
ok, I will factor FS out of this PR.

harshavardhana commented on cmd/api-errors.go:
```diff
@@ -809,19 +810,19 @@ var errorCodes = errorCodeMap{
+	ErrRemoteTargetNotFoundError: {
+		Code:           "XminioAdminRemoteTargetNotFoundError",
+		Description:    "The remote target does not exist",
 		HTTPStatusCode: http.StatusNotFound,
```
These changes you can send as part of a separate PR as well to get it sooner in. Since these are verbatim errors..

harshavardhana commented on cmd/bucket-lifecycle-handlers.go:
```diff
@@ -78,6 +78,11 @@ func (api objectAPIHandlers) PutBucketLifecycleHandler(w http.ResponseWriter, r
+	if err := validateLifecycleTransition(ctx, bucket, bucketLifecycle); err != nil {
+		writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL, guessIsBrowserReq(r))
+		return
+	}
```
```suggestion
	}
	
```

harshavardhana commented on cmd/bucket-lifecycle-handlers.go:
```diff
@@ -78,6 +78,11 @@ func (api objectAPIHandlers) PutBucketLifecycleHandler(w http.ResponseWriter, r
 	}
 
+	// Validate the transition storage ARNs
+	if err := validateLifecycleTransition(ctx, bucket, bucketLifecycle); err != nil {
```
```suggestion
	if err = validateLifecycleTransition(ctx, bucket, bucketLifecycle); err != nil {
```

harshavardhana commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,466 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+	var opts ObjectOptions
+	opts.Versioned = globalBucketVersioningSys.Enabled(bucket)
+	opts.VersionID = objInfo.VersionID
+	opts.Transitioned = true //signal object layer to delete object content while leaving metadata intact.
```
```suggestion
	opts.Transitioned = true // signal object layer to delete object content while leaving metadata intact.
```

harshavardhana commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,466 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+	if err != nil {
+		return err
+	}
+	return nil
```
```suggestion
	_, err = objectAPI.DeleteObject(ctx, bucket, object, opts)
	return err
```

harshavardhana commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,466 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+
+var (
+	errMissingRestoreHdr = fmt.Errorf("x-amz-restore header not found")
+	errInvalidRestoreHdr = fmt.Errorf("x-amz-restore header malformed")
```
```suggestion
	errorRestoreHDRMalformed = fmt.Errorf("x-amz-restore header malformed")
```

harshavardhana replied:
Friendly for editor completions..

harshavardhana commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,466 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+}
+
+var (
+	errMissingRestoreHdr = fmt.Errorf("x-amz-restore header not found")
```
```suggestion
	errRestoreHDRMissing = fmt.Errorf("x-amz-restore header not found")
```

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -642,6 +643,23 @@ func (fs *FSObjects) CopyObject(ctx context.Context, srcBucket, srcObject, dstBu
+			}
+			return oi, nil
+		}
 
```
Adding these to FS mode is overkill IMO we should avoid increasing scope for FS mode.

harshavardhana commented on cmd/fs-v1.go:
```diff
@@ -1331,6 +1383,38 @@ func (fs *FSObjects) DeleteObject(ctx context.Context, bucket, object string, op
+				return objInfo, toObjectErr(err, bucket, object)
+			}
+			return ObjectInfo{Bucket: bucket, Name: object}, nil
+		}
```
No FS mode support would make this PR simpler to review. Let's leave FS mode as is IMO and let the users use erasure coding. Transition into different storage classes only makes sense for erasure coding.

harshavardhana commented on pkg/bucket/lifecycle/transition.go:
```diff
@@ -18,25 +18,151 @@ package lifecycle
+		return errTransitionInvalid
+	}
+	if t.StorageClass != "" {
+		if _, err := madmin.ParseARN(t.StorageClass); err != nil {
```
```suggestion
		if _, err := madmin.ParseARN(t.StorageClass); err != nil {
```
Can ParseARN not return `errXMLNotWellFormed` ?

harshavardhana replied:
can we not use errMalformedXML a much nicer way to return an error. NOTE `api-errors.go` indeed handle `*xml.UnmarshalError` error returned.

harshavardhana replied:
Also this looks like very MinIO specific are you sure we should re-purpose StorageClass as ARNs?  why not have them labeled to something more humanly readable? so that this parsing can be avoided?

poornas replied:
we can add a --label flag in `mc admin bucket remote add` that could be used in `mc ilm add` for storage-class to make it more compatible.

harshavardhana commented on cmd/object-handlers.go:
```diff
@@ -3207,3 +3210,185 @@ func (api objectAPIHandlers) DeleteObjectTaggingHandler(w http.ResponseWriter, r
+
+	restoreObject := mustGetUUID()
+	if rreq.OutputLocation.S3.BucketName != "" {
+		w.Header()[xhttp.AmzRestoreOutputPath] = []string{fmt.Sprintf("%s/%s/%s", rreq.OutputLocation.S3.BucketName, rreq.OutputLocation.S3.Prefix, restoreObject)}
```
isn't `pathJoin` an option here instead of `fmt.Sprintf` ? @poornas

harshavardhana commented on cmd/http/headers.go:
```diff
@@ -66,6 +66,12 @@ const (
+	AmzRestore            = "X-Amz-Restore"
+	AmzRestoreExpiryDays  = "X-Amz-Restore-Expiry-Days"
+	AmzRestoreRequestDate = "X-Amz-Restore-Request-Date"
+	AmzRestoreOutputPath  = "X-Amz-Restore-Output-Path"
```
have you verified if this lower cased in case of AWS s3?

poornas replied:
is there a difference? I see the AWS docs showing lowercase but the output of `aws s3api head-object` returned camel cased which is what I based the header names on.

harshavardhana replied:
We are basically trying to send exactly how AWS might send, i.e on the wire. just to be sure to avoid confusion with poorly implemented SDKs. Ideally, HTTP headers are case insensitive it is to save a conversation.

Commit:
Add support for ILM transition

This PR adds transition support for ILM
to transition data to another MinIO target
represented by a storage class ARN. Subsequent
GET or HEAD for that object will be streamed from
the transition tier. If PostRestoreObject API is
invoked, the transitioned object can be restored for
duration specified to the source cluster.

klauspost commented on cmd/object-api-interface.go:
```diff
@@ -42,9 +42,12 @@ type ObjectOptions struct {
 	PartNumber           int                 // only useful in case of GetObject/HeadObject
 	CheckPrecondFn       CheckPreconditionFn // only set during GetObject/HeadObject/CopyObjectPart preconditional valuation
+	Transitioned         bool                // indicates if the object is to be in transitioned state
+
```
```suggestion
```

klauspost reviewed (approved):
Beautiful!

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,474 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+		// When an object is past expiry, delete the data from transitioned tier and
+		// metadata from source
+		if err := tgt.RemoveObject(ctx, arn.Bucket, object, miniogo.RemoveObjectOptions{VersionID: objInfo.VersionID}); err != nil {
+			logger.LogIf(ctx, err)
```
Just a question, does transition succeed later if we transition an object with the same name in the future ?

poornas replied:
I do not see when this might be an issue - transition metadata is specific to the object or object version.

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,474 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+	opts.VersionID = objInfo.VersionID
+	switch action {
+	case lifecycle.DeleteRestoredAction, lifecycle.DeleteRestoredVersionAction:
+		if time.Now().UTC().After(objInfo.RestoreExpires) {
```
This test in theory should not be needed, we already computed the action previously

poornas replied:
yes, removed this check

vadmeste commented on cmd/bucket-lifecycle.go:
```diff
@@ -46,3 +64,474 @@ func (sys *LifecycleSys) Get(bucketName string) (lc *lifecycle.Lifecycle, err er
+		}
+		// Notify object deleted event.
+		sendEvent(eventArgs{
+			EventName:  event.ObjectRemovedDelete,
```
Sometimes the event is s3:ObjectRemoved:DeleteMarkerCreated

poornas replied:
thanks, took care of this.

vadmeste commented on cmd/data-crawler.go:
```diff
@@ -744,15 +753,27 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
+	case lifecycle.TransitionAction, lifecycle.TransitionVersionAction:
+		if err := transitionObject(ctx, o, i.bucket, i.objectPath(), obj, lcOpts); err != nil {
+			logger.LogIf(ctx, err)
+			return size
```
We return 'size' here because we want the crawler to count the data size as well, but if we are transitioning the object in other places, I am not sure if we need to do that

poornas replied:
we are returning size only if transitioning failed, otherwise it is 0.

vadmeste reviewed (commented):
Some comments, not tested yet

vadmeste commented on cmd/data-crawler.go:
```diff
@@ -744,15 +755,33 @@ func (i *crawlItem) applyActions(ctx context.Context, o ObjectLayer, meta action
+		opts.VersionID = obj.VersionID
+		opts.Transitioned = true // signal object layer to delete object content while leaving metadata intact.
+		opts.TransitionStatus = TransitionPending
+		if _, err = o.DeleteObject(ctx, obj.Bucket, obj.Name, opts); err != nil {
```
How it comes that we delete object content without successfully checking if the transition is properly completed ?

poornas replied:
here - we are not doing an actual Delete, just leveraging the DeleteObject API to mark the object metadata as  "pending" transition. The queuing of transition happens later based on available bandwidth

vadmeste reviewed (commented):
Overall it is stable, but I found two bugs:
* RestoreObject seems to cause an issue when the object expiration time comes, unable to download the object.
* Not sure but it seems that when the transition between site 1 to site B fails, site 1 won't try to transition the object anymore although the cause of the problem disappears (like transition an object version while site B does not have versioning enabled)

kerneltime commented on cmd/admin-bucket-handlers.go:
```diff
@@ -129,16 +129,17 @@ func (a adminAPIHandlers) SetRemoteTargetHandler(w http.ResponseWriter, r *http.
 	vars := mux.Vars(r)
 	bucket := vars["bucket"]
 
+	if !globalIsErasure {
```
Nit: Not sure why access denied is after is the operation possible? If the req does not have admin priv, that should be the first failure.

poornas replied:
we do not want user thinking it is a admin privilege issue if the gateway/fs mode does not have transition support.

kerneltime commented on pkg/madmin/remote-target-commands.go:
```diff
@@ -35,11 +35,13 @@ type ServiceType string
+	ILMService ServiceType = "ilm"
 )
 
 // IsValid returns true if ARN type represents replication
```
Nit: Comment needs updating.

kerneltime commented on cmd/storage-datatypes.go:
```diff
@@ -112,6 +112,13 @@ type FileInfo struct {
+	Transitioned bool
+	// TransitionStatus is set to Pending/Complete for transitioned
+	// entries based on state of transition
+	TransitionStatus string
```
Why is the boolean needed? Is it possible to just use the string?

poornas replied:
with transition queuing, this is no longer needed. Removed it

kerneltime commented on pkg/bucket/lifecycle/lifecycle.go:
```diff
@@ -221,22 +265,43 @@ func (lc Lifecycle) ComputeAction(obj ObjectOpts) Action {
-//        expire in 1 day, then the expected expiry time is `Fri, 23 May 2020 00:00:00 GMT`
-func expectedExpiryTime(modTime time.Time, days ExpirationDays) time.Time {
+//       transition in 1 day, then the expected transition time is `Fri, 23 May 2020 00:00:00 GMT`
+func ExpectedExpiryTime(modTime time.Time, days int) time.Time {
```
Nit: Change days to `Duration` and convert where the api is called?

poornas replied:
just reusing an existing function for both lifecycle and transition - they use typed aliases which would now need to be cast to time.Duration instead making it clunkier to call. Leaving this as is.

harshavardhana commented on cmd/xl-storage.go:
```diff
@@ -1009,7 +1010,7 @@ func (s *xlStorage) DeleteVersion(ctx context.Context, volume, path string, fi F
 
 	// when data-dir is specified.
-	if dataDir != "" {
+	if dataDir != "" && fi.TransitionStatus != lifecycle.TransitionPending {
```
```suggestion
	if dataDir != "" && fi.TransitionStatus != lifecycle.TransitionPending {
```

Need to add some context here what this means

harshavardhana commented on cmd/xl-storage.go:
```diff
@@ -1020,7 +1021,7 @@ func (s *xlStorage) DeleteVersion(ctx context.Context, volume, path string, fi F
 	}
 
-	if !lastVersion {
+	if !lastVersion || fi.TransitionStatus != "" {
```
Add some more context why we are interested in TransitionStatus?

poornas replied:
done

harshavardhana commented on cmd/xl-storage.go:
```diff
@@ -1164,7 +1165,10 @@ func (s *xlStorage) ReadVersion(ctx context.Context, volume, path, versionID str
+	// after transitioning, data dir will be missing
+	if fi.TransitionStatus != "" {
+		checkDataDir = false
+	}
```
Add context why data dir will be missing.

vadmeste reviewed (approved):
LGTM
